staging/src/k8s.io/apiserver/pkg/server/filters/with_not_ready_patch.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"errors"
+	"k8s.io/apiserver/pkg/warning"
+	"net/http"
+
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// WithNotReady rejects any incoming new request(s) with a 429
+// if the specified hasBeenreadyCh channel is closed
+//
+// It includes new request(s) on a new or an existing TCP connection
+// Any new request(s) arriving before hasBeenreadyCh is closed
+// are replied with a 429 and the following response headers:
+//   - 'Retry-After: N` (so client can retry after N seconds, hopefully on a new apiserver instance)
+//
+// if exemptDebuggerGroup is set, all request(s) from the debugger group is exempt,
+// this helps debug the apiserver if it takes longer to initialize.
+func WithNotReady(handler http.Handler, hasBeenReadyCh <-chan struct{}, exemptDebuggerGroup string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		select {
+		case <-hasBeenReadyCh:
+			handler.ServeHTTP(w, req)
+			return
+		default:
+		}
+
+		requestor, exists := request.UserFrom(req.Context())
+		if !exists {
+			responsewriters.InternalError(w, req, errors.New("no user found for request"))
+			return
+		}
+
+		// make sure we exempt:
+		//  - local loopback requests (this exempts system:apiserver)
+		//  - health probes and metric scraping
+		//  - requests from the exempt debugger group.
+		if requestor.GetName() == user.APIServerUser ||
+			hasExemptPathPrefix(req) ||
```

Comment:
we could simplify that by calling [isRequestExemptFromRetryAfter](https://github.com/openshift/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go#L107-L113)

Reply:
technically, we don't need to look at the User-Agent header since we have already checked whether it's local loopback by `requestor.GetName() == user.APIServerUser`